File: Product of Array Except Self/product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        res = [1] * len(nums)
        
        
        prefix = 1
        for i in range(1, len(nums)):
            prefix *= nums[i - 1]
            res[i] = prefix
        
        
        suffix = 1
        for i in range(len(nums) - 2, -1, -1):
            suffix *= nums[i + 1]
            res[i] *= suffix
        
        return res

File: Product of Array Except Self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size(),1);
        int prefix=1;
        for(int i=1;i<nums.size();i++){
            prefix*=nums[i-1];
            res[i]=prefix;
        }
        int suffix=1;
        for(int i=nums.size()-2;i>=0;i--){
            suffix*=nums[i+1];
            res[i]=suffix*res[i];
        }
        return res;
    }
};